sim/midas/src/main/scala/midas/core/CPUManagedStreamEngine.scala

```diff
@@ -0,0 +1,213 @@
+
+// See LICENSE for license details.
+
+package midas.core
+
+import chisel3._
+import chisel3.util._
+import chisel3.experimental.IO
+import freechips.rocketchip.amba.axi4._
+import freechips.rocketchip.config.{Parameters, Field}
+import freechips.rocketchip.diplomacy._
+import freechips.rocketchip.util.DecoupledHelper
+
+import midas.widgets._
+import midas.widgets.CppGenerationUtils._
+
+/**
+  * A helper container to serialize per-stream constants to the header. This is
+  * currently somewhat redundant with the default header emission for widgets.
+  */
+case class StreamDriverParameters(
+  channelName: String,
+  bufferBaseAddress: Int,
+  mmioRegisters: (String, Int)*) {
+
+  def headerFragment(mmioBaseAddress: BigInt): String = {
+    val registerMacros = for ((name, offset) <- mmioRegisters) yield {
+      genConstStatic(s"${channelName}_${name}_address", UInt32(mmioBaseAddress + offset))
+    }
+    val dmaAddressMacro = genConstStatic(s"${channelName}_dma_address", UInt32(bufferBaseAddress))
+
+    (dmaAddressMacro +: registerMacros.toSeq).mkString
+  }
+}
+
+class CPUManagedStreamEngine(p: Parameters, val params: StreamEngineParameters) extends StreamEngine(p) {
+
+  val dmaBytes = p(DMANastiKey).dataBits / 8
+  val pcisNodeOpt = Some(AXI4SlaveNode(
+      Seq(AXI4SlavePortParameters(
+        slaves = Seq(AXI4SlaveParameters(
+          address       = Seq(AddressSet(0, (BigInt(1) << p(DMANastiKey).dataBits) - 1)),
+          resources     = (new MemoryDevice).reg,
+          regionType    = RegionType.UNCACHED, // cacheable
+          executable    = false,
+          supportsWrite = TransferSizes(dmaBytes, 4096),
+          supportsRead  = TransferSizes(dmaBytes, 4096),
+          interleavedId = Some(0))), // slave does not interleave read responses
+        beatBytes = dmaBytes)
+    ))
+  )
+
+  //require(BridgeStreamConstants.streamWidth == p(DMANastiKey).dataBits,
+  //  s"CPU-mastered AXI4 IF data width must match the stream width ${BridgeStreamConstants.streamWidth}".)
+
+  val pcimNodeOpt = None
+
+  lazy val module = new WidgetImp(this) {
+    val io = IO(new WidgetIO)
+
+    val dma = pcisNodeOpt.get.in.head._1
+
+    // FromHostCPU streams are implemented using the AW, W, B channels, which
+    // write into large BRAM FIFOs for each stream.
+    assert(!dma.aw.valid || dma.aw.bits.size === log2Ceil(dmaBytes).U)
+    assert(!dma.w.valid  || dma.w.bits.strb === ~0.U(dmaBytes.W))
+
+    dma.b.bits.resp := 0.U(2.W)
+    dma.b.bits.id := dma.aw.bits.id
+    dma.b.bits.user := dma.aw.bits.user
+    // This will be set by the channel given the grant using last connect semantics
+    dma.b.valid := false.B
+    dma.aw.ready := false.B
+    dma.w.ready := false.B
+
+
+    // TODO: Chisel naming prefix to indicate what channel this hw belongs to.
+    // This demultiplexes the AW, W, and B channels onto the decoupled ports representing each stream.
+    def elaborateFromHostCPUStream(
```

Comment:
The bodies of these functions are more or less copied directly from the old Implementation.